Mathlib/Data/Prod/Basic.lean

```diff
@@ -0,0 +1,275 @@
+
+/-!
+# Extra facts about `prod`
+
+This file defines `prod.swap : α × β → β × α` and proves various simple lemmas about `prod`.
+-/
+
+variable {α : Type _} {β : Type _} {γ : Type _} {δ : Type _}
+
+@[simp]
+theorem Prod_map (f : α → γ) (g : β → δ) (p : α × β) : Prod.map f g p = (f p.1, g p.2) :=
+  rfl
+#align prod_map Prod_map
+
+namespace Prod
+
+@[simp]
+theorem «forall» {p : α × β → Prop} : (∀ x, p x) ↔ ∀ a b, p (a, b) :=
+  ⟨fun h a b => h (a, b), fun h ⟨a, b⟩ => h a b⟩
+
+@[simp]
+theorem «exists» {p : α × β → Prop} : (∃ x, p x) ↔ ∃ a b, p (a, b) :=
+  ⟨fun ⟨⟨a, b⟩, h⟩ => ⟨a, b, h⟩, fun ⟨a, b, h⟩ => ⟨⟨a, b⟩, h⟩⟩
+
+theorem forall' {p : α → β → Prop} : (∀ x : α × β, p x.1 x.2) ↔ ∀ a b, p a b :=
+  Prod.forall
+
+theorem exists' {p : α → β → Prop} : (∃ x : α × β, p x.1 x.2) ↔ ∃ a b, p a b :=
+  Prod.exists
+
+@[simp]
+theorem snd_comp_mk (x : α) : Prod.snd ∘ (Prod.mk x : β → α × β) = id :=
+  rfl
+
+@[simp]
+theorem fst_comp_mk (x : α) : Prod.fst ∘ (Prod.mk x : β → α × β) = Function.const β x :=
+  rfl
+
+@[simp]
+theorem map_mk (f : α → γ) (g : β → δ) (a : α) (b : β) : map f g (a, b) = (f a, g b) :=
+  rfl
+
+theorem map_fst (f : α → γ) (g : β → δ) (p : α × β) : (map f g p).1 = f p.1 :=
+  rfl
+
+theorem map_snd (f : α → γ) (g : β → δ) (p : α × β) : (map f g p).2 = g p.2 :=
+  rfl
+
+theorem map_fst' (f : α → γ) (g : β → δ) : Prod.fst ∘ map f g = f ∘ Prod.fst :=
+  funext <| map_fst f g
+
+theorem map_snd' (f : α → γ) (g : β → δ) : Prod.snd ∘ map f g = g ∘ Prod.snd :=
+  funext <| map_snd f g
+
+/-- Composing a `Prod.map` with another `Prod.map` is equal to
+a single `Prod.map` of composed functions.
+-/
+theorem map_comp_map {ε ζ : Type _} (f : α → β) (f' : γ → δ) (g : β → ε) (g' : δ → ζ) :
+    Prod.map g g' ∘ Prod.map f f' = Prod.map (g ∘ f) (g' ∘ f') :=
+  rfl
+
+/-- Composing a `Prod.map` with another `Prod.map` is equal to
+a single `Prod.map` of composed functions, fully applied.
+-/
+theorem map_map {ε ζ : Type _} (f : α → β) (f' : γ → δ) (g : β → ε) (g' : δ → ζ) (x : α × γ) :
+    Prod.map g g' (Prod.map f f' x) = Prod.map (g ∘ f) (g' ∘ f') x :=
+  rfl
+
+-- Porting note: mathlib3 proof uses `by cc` for the mpr direction
+-- Porting note: `@[simp]` tag removed because auto-generated `mk.injEq` simplifies LHS
+-- @[simp]
+theorem mk.inj_iff {a₁ a₂ : α} {b₁ b₂ : β} : (a₁, b₁) = (a₂, b₂) ↔ a₁ = a₂ ∧ b₁ = b₂ :=
+  Iff.of_eq (mk.injEq _ _ _ _)
+
+theorem mk.inj_left {α β : Type _} (a : α) : Function.Injective (Prod.mk a : β → α × β) := by
+  intro b₁ b₂ h
+  simpa only [true_and, Prod.mk.inj_iff, eq_self_iff_true] using h
+
+theorem mk.inj_right {α β : Type _} (b : β) :
+    Function.Injective (fun a => Prod.mk a b : α → α × β) := by
+  intro b₁ b₂ h
+  simpa only [and_true, eq_self_iff_true, mk.inj_iff] using h
+
+theorem ext_iff {p q : α × β} : p = q ↔ p.1 = q.1 ∧ p.2 = q.2 := by
+  rw [← @mk.eta _ _ p, ← @mk.eta _ _ q, mk.inj_iff]
+
+@[ext]
+theorem extₓ {α β} {p q : α × β} (h₁ : p.1 = q.1) (h₂ : p.2 = q.2) : p = q :=
```

Comment:
I don't think we should be using `ₓ` in names! This is/was just an internal mathport thing.

Reply:
Would you prefer the previous `ext'` name?

Reply:
Yes.